include/nbl/video/SPhysicalDeviceLimits.h

```diff
@@ -0,0 +1,497 @@
+#ifndef _NBL_VIDEO_S_PHYSICAL_DEVICE_LIMITS_H_INCLUDED_
+#define _NBL_VIDEO_S_PHYSICAL_DEVICE_LIMITS_H_INCLUDED_
+
+#include <type_traits>
+#include "nbl/asset/utils/IGLSLCompiler.h" // asset::IGLSLCompiler::E_SPIRV_VERSION
+
+
+namespace nbl::video
+{
+
+struct SPhysicalDeviceLimits
+{
+    /* Vulkan 1.0 Core  */
+    uint32_t maxImageDimension1D;
+    uint32_t maxImageDimension2D;
+    uint32_t maxImageDimension3D;
+    uint32_t maxImageDimensionCube;
+    uint32_t maxImageArrayLayers;
+    uint32_t maxBufferViewSizeTexels;
+    uint32_t maxUBOSize;
+    uint32_t maxSSBOSize;
+    //uint32_t              maxPushConstantsSize;
+    //uint32_t              maxMemoryAllocationCount;
+    //uint32_t              maxSamplerAllocationCount;
+    //VkDeviceSize          bufferImageGranularity;
+    //VkDeviceSize          sparseAddressSpaceSize;
+    //uint32_t              maxBoundDescriptorSets;
+    //uint32_t              maxPerStageDescriptorSamplers;
+    //uint32_t              maxPerStageDescriptorUniformBuffers;
+    uint32_t maxPerStageDescriptorSSBOs;
+    //uint32_t              maxPerStageDescriptorSampledImages;
+    //uint32_t              maxPerStageDescriptorStorageImages;
+    //uint32_t              maxPerStageDescriptorInputAttachments;
+    //uint32_t              maxPerStageResources;
+    //uint32_t              maxDescriptorSetSamplers;
+    uint32_t maxDescriptorSetUBOs;
+    uint32_t maxDescriptorSetDynamicOffsetUBOs;
+    uint32_t maxDescriptorSetSSBOs;
+    uint32_t maxDescriptorSetDynamicOffsetSSBOs;
+    uint32_t maxDescriptorSetImages;
+    uint32_t maxDescriptorSetStorageImages;
+    //uint32_t              maxDescriptorSetInputAttachments;
+    //uint32_t              maxVertexInputAttributes;
+    //uint32_t              maxVertexInputBindings;
+    //uint32_t              maxVertexInputAttributeOffset;
+    //uint32_t              maxVertexInputBindingStride;
```

Comment:
These seem important too, when constructing vertex buffers and attributes to give to pipeline

Reply:
dont uncomment any of them, let me explain why

### maxVertexInputAttributes and maxVertexInputBindings

In OpenGL (and ES) the de-jure (legal) minimum is 16, and de-facto (in practice) Vulkan reports begin at 16.

### maxVertexInputAttributeOffset and maxVertexInputBindingStride

In OpenGL (and ES) the de-jure (legal) minimum is 2047 for both, and de-facto (in practice) Vulkan reports begin at 2047.

## Assets Serialization

An `ICPUMeshBuffer` is an `IAsset` and for reasons of serialization and conversion we've hardcoded the attribute and binding count to 16 (the bitfields, array sizes, etc.), variable attribute count meshes would be a mess.

Anyway its all futile effort to split hairs there in the age of programmable pulling, visibility buffer, bindless vertices and flexible vertex indexing. I've never needed to use more than 16 vertex attributes, except when instancing and pushing MVP and MV per-instance matrices.

Furthermore you wouldn't want to be able to create a CPU asset which cannot be easily converted to a GPU resource because it has too many vertex attributes or bindings.

Format promotion is bad enough, and would require repacking (due to bad offsets/strides) is easy, solving surplus attributes and bindings isn't (when repacking you potentially need to use 1 binding for each 1 attribute).

Reply:
obviously put that justification in the comments